include/boost/hana/index_if.hpp

```diff
@@ -0,0 +1,114 @@
+/*!
+@file
+Defines `boost::hana::index_if`.
+
+@copyright Louis Dionne 2013-2017
+@copyright Jason Rice 2017
+Distributed under the Boost Software License, Version 1.0.
+(See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_HANA_INDEX_IF_HPP
+#define BOOST_HANA_INDEX_IF_HPP
+
+#include <boost/hana/concept/foldable.hpp>
+#include <boost/hana/concept/iterable.hpp>
+#include <boost/hana/config.hpp>
+#include <boost/hana/detail/decay.hpp>
+#include <boost/hana/detail/index_if.hpp>
+#include <boost/hana/fwd/at.hpp>
+#include <boost/hana/fwd/basic_tuple.hpp>
+#include <boost/hana/fwd/index_if.hpp>
+#include <boost/hana/integral_constant.hpp>
+#include <boost/hana/length.hpp>
+#include <boost/hana/optional.hpp>
+
+#include <cstddef>
+#include <utility>
+
+
+BOOST_HANA_NAMESPACE_BEGIN
+    //! @cond
+    template <typename Xs, typename Pred>
+    constexpr auto index_if_t::operator()(Xs&& xs, Pred&& pred) const {
+        using S = typename hana::tag_of<Xs>::type;
+        using IndexIf = BOOST_HANA_DISPATCH_IF(index_if_impl<S>,
+            hana::Iterable<S>::value
+        );
+
+    #ifndef BOOST_HANA_CONFIG_DISABLE_CONCEPT_CHECKS
+        static_assert(hana::Iterable<S>::value,
+        "hana::index_if(xs, pred) requires 'xs' to be a Iterable");
+    #endif
+
+        return IndexIf::apply(static_cast<Xs&&>(xs), static_cast<Pred&&>(pred));
+    }
+    //! @endcond
+
+    namespace detail {
+        template <std::size_t i, std::size_t N, bool Done>
+        struct iterate_while;
+
+        template <std::size_t i, std::size_t N>
+        struct iterate_while<i, N, false> {
+            template <typename F, typename Pred>
+            using f = typename iterate_while<i + 1, N,
```

Comment:
We could be more compile-time efficient here if we hard-coded the `hana::at` call, since we would avoid instantiating `partial_at::operator()` on each iteration. Unless we reuse `iterate_while` somewhere else with a different `F`, I suggest we do that